Documentation for phase shift transformer (PST) (#1491)
PowerGridModel/power-grid-model: docs/user_manual/components.md

##### Mapping from CGMES / ENTSO-E PST models
<!-- markdownlint-disable-next-line MD013 -->
The [ENTSO-E PST modelling document (CGMES v2.4)](https://eepublicdownloads.entsoe.eu/clean-documents/CIM_documents/Grid_Model_CIM/ENTSOE_CGMES_v2.4_28May2014_PSTmodelling.pdf)
describes the standard PST types and their mapping to CIM classes.
All of these types can be represented with `generic_branch`.
In the ENTSO-E models the resistances and magnetizing currents are neglected, which corresponds to
`r1 = 0`, `g1 = 0`, `b1 = 0`; only `x1`, `k` and `theta` remain.
With $m = n - n_0$ (CIM: `step` − `neutralStep`), $\Delta u$ = `voltageStepIncrement` and
$\psi$ = `windingConnectionAngle`, the per-tap values are:

| CIM class                                | angle $\alpha$ per tap                                                                       | magnitude ratio                                                                   |
| ---------------------------------------- | -------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------- |

> | `PhaseTapChangerLinear`                  | $\alpha = m \cdot \delta$ (`stepPhaseShiftIncrement`)                                        | constant                                                                          |

Reviewer: nitpicks:

```suggestion
| `PhaseTapChangerLinear`                  | $\alpha = m \cdot \delta$    \theta_{\text{step}}                                         | constant (k_{\text{off-nominal}})                                     |
```
By constant, it would then be equal to off nominal tap and right?
Also can you add `stepPhaseShiftIncrement = theta_step` above for consistency.

Reviewer 2: Thanks @nitbharambe — to answer the question itself: yes, your reading is
correct. For `PhaseTapChangerLinear` only the angle scales with the tap step;
the magnitude ratio stays fixed at the configured off-nominal ratio, i.e. the
same `k` as in the generic_branch model (`N = k · e^{jθ}`), independent of tap
position. That is exactly what distinguishes it from `PhaseTapChangerAsymmetrical`.

Your naming suggestion makes sense. As I noted above, with the content now
living in `non-pgm-components.md` via #1485 I don't see much left for this PR
to add, so please go ahead and apply the rename directly there — I'll leave it
in your hands. Two things to watch when you do: `\theta_{\text{step}}` and
`k_{\text{off-nominal}}` need their `$` delimiters to render, and
`stepPhaseShiftIncrement = θ_step` should go into the symbol legend above the
table.

Reviewer 3: Resolved as [38bda9e](https://github.com/PowerGridModel/power-grid-model/pull/1491/commits/38bda9e308385cf633fc6571d40085a36b9b106f)

\end{aligned}
$$

#### Modeling a phase-shifting transformer

A phase-shifting transformer (PST) can be modeled directly with `generic_branch`.
The complex ratio

$$
N = k \cdot e^{\mathrm{j} \theta}
$$

already contains the phase shift; the series impedance and the magnetizing branch are given by

> `r1`, `x1`, `g1`, `b1` as for any transformer. No dedicated PST component is required.

Reviewer 3: ```suggestion
already contains the phase shift; the series impedance (`r1` and `x1`) and the magnetizing branch (`g1` and `b1`) are given as input for any transformer. No dedicated PST component is needed.
```

> The angle `theta` must be provided explicitly by the user.

Reviewer 3: ```suggestion
Note that the angle `theta` must be provided explicitly by user.
```

For a transformer with a tap changer, `theta` (and in general also `k` and `x1`, see below) has to
be calculated from the tap position and the transformer data, i.e. from the additional (boost)

> voltage injected per tap step and its injection angle.

Reviewer 3: ```suggestion
For transformers with tap changers, the parameters `k`, `theta`, and `x1` are dynamic.
They must be calculated based on the current tap position, using the voltage magnitude boost and
the phase injection angle defined per tap step.
```

Let

* $m = \text{tap\_pos} - \text{tap\_neutral}$ be the tap position relative to the neutral position,
* $\Delta u$ be the additional voltage per tap step in per-unit of the winding voltage,
* $\psi$ be the injection angle of the additional voltage relative to the winding voltage

>   (e.g. $\psi = 90^{\circ}$ for pure quadrature/phase regulation).

Reviewer 3: ```suggestion
  (e.g., $\psi = 90^{\circ}$ for pure quadrature/phase regulation).
```
minor punctuation tweak to match the rest of the repo doc.

The tap-dependent complex ratio is then

$$
\rho = 1 + m \, \Delta u \, e^{\mathrm{j} \psi}
$$

so that

$$
\begin{aligned}
k_{\text{tap}} &= |\rho| = \sqrt{\left(1 + m \, \Delta u \cos\psi\right)^2
                                 + \left(m \, \Delta u \sin\psi\right)^2} \\
\theta_{\text{tap}} &= \operatorname{atan2}\!\left(m \, \Delta u \sin\psi,\;
                                                  1 + m \, \Delta u \cos\psi\right)
\end{aligned}
$$

> The `generic_branch` input attributes follow as

Reviewer 3: ```suggestion
The `generic_branch` input attributes `k` and `theta` follow as
```

$$
\begin{aligned}
k      &= k_{\text{off-nominal}} \cdot k_{\text{tap}} \\
\theta &= \theta_{\text{fixed}} + \theta_{\text{tap}}
\end{aligned}
$$

where $k_{\text{off-nominal}}$ is the off-nominal ratio of the transformer at the neutral tap
position and $\theta_{\text{fixed}}$ is the fixed phase shift of the winding configuration

> (e.g. from the clock number).

Reviewer 3: ```suggestion
(e.g., from the clock number).
```

**Special case: ideal phase-shifting tap changer.**

> Some tools (e.g. pandapower with `tap_changer_type="Ideal"`) model a tap changer that changes only

Reviewer 3: ```suggestion
Some tools (e.g., pandapower with `tap_changer_type="Ideal"`) model tap changers that change only
```

the phase angle, while the voltage magnitude ratio stays constant.
Note that "ideal" refers to the tap changer, not to a zero-impedance transformer: the series
impedance and the magnetizing branch are still modeled through `r1`, `x1`, `g1`, `b1`.

> In this case the attributes reduce to

Reviewer 3: ```suggestion
In this case, `k` and `theta` reduce to
```

$$
\begin{aligned}
k      &= k_{\text{off-nominal}} \\
\theta &= \theta_{\text{fixed}} + m \cdot \theta_{\text{step}}
\end{aligned}
$$

with $\theta_{\text{step}}$ the angle shift per tap step

> (pandapower: `tap_step_degree`, converted to radian).

Reviewer 3: ```suggestion
(equivalent to `tap_step_degree` in pandapower: converted to radian).
```

A lossless but not impedance-free PST is obtained with `r1 = 0`, `x1 > 0`, `g1 = 0`, `b1 = 0`.

> ##### Mapping from CGMES / ENTSO-E PST models

Reviewer 3: This section follows a forward presentation of the mapping. I wonder whether it makes sense to reduce the logic flow and present the mapping directly.

Reviewer 3: Proceed as a full reference in its current state. Resolved.

<!-- markdownlint-disable-next-line MD013 -->
The [ENTSO-E PST modelling document (CGMES v2.4)](https://eepublicdownloads.entsoe.eu/clean-documents/CIM_documents/Grid_Model_CIM/ENTSOE_CGMES_v2.4_28May2014_PSTmodelling.pdf)
describes the standard PST types and their mapping to CIM classes.
All of these types can be represented with `generic_branch`.
In the ENTSO-E models the resistances and magnetizing currents are neglected, which corresponds to

> `r1 = 0`, `g1 = 0`, `b1 = 0`; only `x1`, `k` and `theta` remain.

Reviewer 3: ```suggestion
In ENTSO-E models, the windings resistance and core magnetizing components are neglected (`r1 = 0`, `g1 = 0`, `b1 = 0`);
consequently, the transformer is defined solely by its reactance (`x1`), tap ratio (`k`), and phase shift angle (`theta`).
```

> With $m = n - n_0$ (CIM: `step` − `neutralStep`), $\Delta u$ = `voltageStepIncrement` and

Reviewer 3: ```suggestion
With $m = n - n_0$ (in CIM: `step` − `neutralStep`), $\Delta u$ = `voltageStepIncrement` and
```

\end{aligned}
$$

For `generic_branch` this means that a tap change on a CGMES PST in general requires recalculating
**three** attributes: `k`, `theta` and `x1`.
When `PhaseTapChangerTabular` data is available (recommended by ENTSO-E), `k`, `theta` and `x1` can
be taken directly from the per-tap table columns `ratio`, `angle` and `x` without evaluating the
formulas above.

```{note}
Mind the direction convention: the ENTSO-E document defines the ratio as
$r = 1 / \lvert\rho\rvert$ (referenced from the opposite side).
Depending on which side of the `generic_branch` the tap winding is located, either
$\lvert\rho\rvert$ or its reciprocal must be used for `k`, consistent with the fact that `r1`,
`x1`, `g1`, `b1` are referenced to the "to" side of the branch.
```

```{warning}
A fully impedance-free ideal phase link (`r1 = x1 = 0`) cannot be modeled, because the series
admittance $Y_{\text{series}} = 1 / (r_1 + \mathrm{j} x_1)$ would become singular.
```

```{note}
The attributes `k`, `theta` and `x1` are not updatable.
Changing the tap position of a phase-shifting transformer therefore requires recalculating these
attributes and rebuilding the model.

> Automatic tap control via the transformer tap regulator is not available for `generic_branch`.

Reviewer 3: ```suggestion
```
```{note}
Automatic tap control via the transformer tap regulator is not available for `generic_branch`.
```

Reviewer 3: This sentence deserves its own ```{note} was the original comment. Github renders that away.

Reviewer 3: Resolved in [38bda9e](https://github.com/PowerGridModel/power-grid-model/pull/1491/commits/38bda9e308385cf633fc6571d40085a36b9b106f)

```

> An example of the usage of `generic_branch` is given in

Reviewer 3: ```suggestion
A use case of `generic_branch` as PST can be found in
```
